### PR TITLE
feat!: add generate keypair helper

### PR DESCRIPTION
**Pull request by rflechtner**
## fixes https://github.com/KILTprotocol/ticket/issues/3530

Adds and exposes a generateKeypair function that returns a multikey-encoded keypair.
I had to refactor multikey en-and decoding for that.

## How to test:

`generateKeypair` is being used on did helper integration tests.

## Checklist:

- [x] I have verified that the code works
- [x] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [x] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [ ] I have documented the changes (where applicable)
    * Either PR or Ticket to update [the Docs](https://github.com/KILTprotocol/docs)
    * Link the PR/Ticket here

### CODE REVIEW

File: packages/utils/src/Multikey.ts [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  'publicKey'
+> & {
+  secretKey: Uint8Array
+  type: KeyTypeString
+}
+export function decodeMultibaseKeypair(
+  keyPairPublicKey: MultibasePublicKey
+): Pick<KeyringPair, 'publicKey'> & { type: KeyTypeString }
+/**
+ * Decode a Multikey representation of a verification method into its fundamental components: the public key and the key type.
```
Comment (abdulmth) [inline packages/utils/src/Multikey.ts:132]: we probably should add that only base 58 is supported? same for encode.

**Review by abdulmth — APPROVED** [review]

File: packages/utils/src/Multikey.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  MultibaseSecretKey,
+} from '@kiltprotocol/types'
+import { u8aConcat } from '@polkadot/util'
+import { base58Encode } from '@polkadot/util-crypto'
+import { DidError } from './SDKErrors.js'
+
+const MULTICODEC_SECP256K1_PREFIXES = [0xe7, 0x1301] as const
+const MULTICODEC_X25519_PREFIXES = [0xec, 0x1302] as const
+const MULTICODEC_ED25519_PREFIXES = [0xed, 0x1300] as const
+const MULTICODEC_SR25519_PREFIXES = [0xef, 0x1303] as const
```
[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/utils/src/Multikey.ts:26]: public- and secret key prefix. these are different according to the table

File: packages/did/src/Did.utils.ts [hunk 3]
```diff
@@ -161,68 +136,19 @@ const multicodecReversePrefixes: Record<DidVerificationMethodType, number> = {
...
-  if (keyType === undefined) {
-    throw new SDKErrors.DidError(
-      `Cannot decode key type for multibase key "${publicKeyMultibase}".`
-    )
-  }
-  if (keyBytes.length !== expectedPublicKeyLength) {
+  const { publicKey, type } = Multikey.decodeMultibaseKeypair({
+    publicKeyMultibase,
+  })
+  const expectedPublicKeyLength = type === 'secp256k1' ? 33 : 32
```
[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/did/src/Did.utils.ts:142]: makes sense

[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/did/src/Did.utils.ts]: fb44a82c96317a2bd142fa8499bf3ab90bef6265

File: packages/utils/src/Multikey.ts [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    secretKey: Uint8Array
+  }
+): MultibaseKeyPair
+export function encodeMultibaseKeypair(
+  args: Pick<KeyringPair, 'publicKey'> & {
+    type: KnownTypeString
+  }
+): MultibasePublicKey
+/**
+ * Calculate the Multikey representation of a keypair given its type and public/secret keys.
```
[the remark this answers is not in the document]
Comment (rflechtner) [reply packages/utils/src/Multikey.ts]: d69481d40bf17e19c54429c89df106f983c9919b ; I have then moved docstrings to the overloads(1bdf0d8ca39040dfe041f7839520cdaa52d46c76), tailoring them to the respective parameters set

File: packages/utils/src/Multikey.ts (hunk 1 above)
Comment (rflechtner) [reply packages/utils/src/Multikey.ts]: d69481d40bf17e19c54429c89df106f983c9919b ; I have then moved docstrings to the overloads(1bdf0d8ca39040dfe041f7839520cdaa52d46c76), tailoring them to the respective parameters set

**Review by abdulmth — APPROVED** [review]

**Pull request merged**